### PR TITLE
Packing patch

### PR DESCRIPTION
**Pull request by SivilTaram**
Fix the issue of packing - which should be compatible with the `first_eos_indices`

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -69,46 +70,88 @@ def get_args():
...
-    # This relies on tokenizer.eos_token_id == tokenizer.pad_token_id
-    first_eos_indices = [
-        model_inputs["input_ids"][i].index(tokenizer.eos_token_id) 
-        if model_inputs["input_ids"][i][-1] == tokenizer.eos_token_id else args.max_input_length
-        for i in range(len(model_inputs["input_ids"]))
-    ]
+        assert tokenizer.eos_token_id == tokenizer.pad_token_id
+        model_inputs = tokenizer(model_inputs,
+                                 max_length=args.max_input_length,
+                                 # no padding to calculate the real length
```
Comment (Muennighoff) [inline train.py]:
```suggestion
                                 max_length=args.max_input_length,
```

File: train.py [hunk 2]
```diff
@@ -69,46 +70,88 @@ def get_args():
...
+            model_inputs["input_ids"][i].index(tokenizer.eos_token_id)
+            if model_inputs["input_ids"][i][-1] == tokenizer.eos_token_id else args.max_input_length
+            for i in range(len(model_inputs["input_ids"]))
+        ]
 
     if args.compute_loss_on_input:
         assert args.compute_loss_on_instruction is False
         model_inputs["labels"] = [
-            inp[:first_eos_indices[i] + 1] + [-100] * (args.max_input_length - first_eos_indices[i] - 1)
+            list(inp[:first_eos_indices[i] + 1]) + [-100] * (args.max_input_length - first_eos_indices[i] - 1)
```
Comment (Muennighoff) [inline train.py:134]: Why was this necessary? Shouldn't it already be a list?

File: train.py [hunk 3]
```diff
@@ -69,46 +70,88 @@ def get_args():
...
+        padded_attention_masks = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        for i, example_input_ids in enumerate(packed_ids):
+            padded_input_ids[i, :len(example_input_ids)] = example_input_ids
+            # obtain the attention mask
+            attention_mask = create_flatten_attention_mask(packed_lengths[i], args.max_input_length)
+            padded_attention_masks[i] = attention_mask
+        model_inputs["input_ids"] = padded_input_ids
+        model_inputs["attention_mask"] = padded_attention_masks
+        # take the last valid eos token as the end of the sequence
+        first_eos_indices = [len(packed_ids[i]) - 1 for i in range(len(packed_ids))]
```
Comment (Muennighoff) [inline train.py]: I think you can use `packed_lengths` here

File: train.py [hunk 4]
```diff
@@ -69,46 +70,88 @@ def get_args():
...
+        model_inputs = tokenizer(model_inputs,
+                                 max_length=args.max_input_length,
+                                 # no padding to calculate the real length
+                                 padding=False,
+                                 truncation=True)
+        # packing in the input_ids level
+        max_window = args.max_input_length
+        # packed_length maintains a list, each of which is a list of all packed sequence lengths in this sequence
+        packed_lengths, packed_ids,  = [], []
+        temp_lengths, temp_ids = [], []
```
Comment (Muennighoff) [inline train.py]: I think temp_lengths can just be an int that's initialized to 0 rather than a list; Would save the sum( ) below

File: train.py [hunk 5]
```diff
@@ -69,46 +70,88 @@ def get_args():
...
+            packed_lengths.append(temp_lengths.copy())
+            packed_ids.append(temp_ids.copy())
+        # pad the packed_ids into max_window
+        padded_input_ids = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        padded_attention_masks = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        for i, example_input_ids in enumerate(packed_ids):
+            padded_input_ids[i, :len(example_input_ids)] = example_input_ids
+            # obtain the attention mask
+            attention_mask = create_flatten_attention_mask(packed_lengths[i], args.max_input_length)
+            padded_attention_masks[i] = attention_mask
```
Comment (Muennighoff) [inline train.py]:
```suggestion
            padded_attention_masks[i, :sum(packed_lengths[i])] += 1
```

If we make `packed_lengths` just store the total length instead, we can also save the sum

File: train.py (hunk 2 above)
Comment (SivilTaram) [reply train.py:134]: In the case of `data_packing`, it is a part of the numpy array. So we should use list to convert it into a list.
